### PR TITLE
extraction and endpoint fixes

### PR DESCRIPTION
**Pull request by frankschae**
This seems to fix the slightly wrong gradients for `InterpolatingAdjoint` with an inplace `NoiseWrapper`

This part:
```julia
  if t isa Union{Rational,Integer}
      i = searchsortedfirst(W.t,t)
    else
      i = searchsortedfirst(W.t,t-eps(typeof(t)))
    end
```
was missing such that the correct index was not found.

This part 
```julia
 out1 .= W.curW + W.dW
    if W.Z != nothing
      out2 .= W.curZ + W.dZ
    end
```
 is necessary because by construction, `curW` is substracted from `out1` to obtain `W.dW`
https://github.com/SciML/DiffEqNoiseProcess.jl/blob/d6e1d61c4e7d162b58fd91c3c9dcafab802664a9/src/noise_interfaces/noise_wrapper_interface.jl#L24 (That was the hard one..)

### CODE REVIEW

File: src/noise_interfaces/noise_process_interface.jl [hunk 1]
```diff
@@ -480,8 +480,17 @@ end
         push!(W.Z,copy(out2))
       end
     end
+    out1 .= W.curW + W.dW
```
Comment (ChrisRackauckas) [inline src/noise_interfaces/noise_process_interface.jl:483]: so lines 471 and 474 are incorrect and should be deleted?

Comment (ChrisRackauckas) [inline src/noise_interfaces/noise_process_interface.jl:483]:
```suggestion
    @. out1 = W.curW + W.dW
```

File: src/noise_interfaces/noise_process_interface.jl [hunk 2]
```diff
@@ -480,8 +480,17 @@ end
...
+    if W.Z != nothing
+      out2 .= W.curZ + W.dZ
```
Comment (ChrisRackauckas) [inline src/noise_interfaces/noise_process_interface.jl:485]:
```suggestion
      @. out2 = W.curZ + W.dZ
```

**Comment by ChrisRackauckas** [comment]
This should definitely get a test. This is the kind of bug you never want to show up again!

File: src/noise_interfaces/noise_process_interface.jl (hunk 1 above)
Comment (frankschae) [reply src/noise_interfaces/noise_process_interface.jl:483]: is it? I thought for `save_everystep` it should be there

**Comment by frankschae** [comment]
Yeah, this was a nasty one..  I'll try to simplify a bit the current example and then add it as a test.

**Comment by ChrisRackauckas** [comment]
The test failures here look real.

**Comment by frankschae** [comment]
yes, they do.. I hope to fix it today.. They failed because of:

```Julia
 @. out1 = W.curW + W.dW
 if W.Z != nothing
    @.  out2 = W.curZ + W.dZ
 end
```
Maybe I need to check for the size of `dt` .

**Comment by frankschae** [comment]
Could we maybe use a saving Callback here to skip the last `perform_step!()`/ don't save its output if `t> tspan[2]`?

**Comment by ChrisRackauckas** [comment]
>if t> tspan[2]?

Wait, how does that occur?

**Comment by frankschae** [comment]
Sorry, this was wrong: So it is like:

```julia 
checkWrapper = retcode: Success
Interpolation: 1st order linear
t: [0.0999, 0.09999999999999817, 0.1]
u: [[0.656276518579838, 0.38193851982038995], [0.6550957136392241, 0.38111632942571044], [0.6540132361803782, 0.3803494232048912]]
```

and the "good endpoint" would have been reached at `0.09999999999999817` 

```
sol.u[end] = [0.6550957136392345, 0.3811163294256931]
```

**Comment by ChrisRackauckas** [comment]
Oh I see what's going on. 

https://github.com/SciML/StochasticDiffEq.jl/blob/master/src/integrators/integrator_utils.jl#L153

Those tiny steps introduce large numerical errors, so what's done is that, if you step close to a tstop, then time is adjusted. This comes from the weird fact that `t + dt - dt != t` for some numbers. This occurs because when a `tstop` is in the range of the `dt`, then `dt` is changed to satisfy `dt = tstop - t`, but then `t + dt != tstop` and can be approximately epsilon smaller. (The final time point is just handled as the final tstop)

However, the noise term's `t` isn't adjusted, so it requires another small extra step, which is then saved using that final timestep save correction. However, instead what should be happening is that the `curt` of the `dW` (and `dP`) should be fixed to check for the distance from the tstop.

**Comment by frankschae** [comment]
oh that makes sense, I was already wondering why I see just two noise values stored in the final solution.

**Comment by frankschae** [comment]
I'll check again in InterpolatingAdjoint **when** the original problem in 

```julia
interpolate!(..)
if sign(W.dt)*t > sign(W.dt)*W.t[end] || (sign(W.dt)*t < sign(W.dt)*W.t[1] && reverse)```
 ..
end
```
occurs. 

(Without any changes on StochasticDiffEq) even the last test set passes. There, I purposely chose ` dt - 10eps` and stop values such that in each step `dt` of the integrator and `W.dt` had to be changed properly.

**Pull request merged**